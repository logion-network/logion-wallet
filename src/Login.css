.Login {
    height: 100vh;
    background: linear-gradient(0deg, #9f1278 0%, #3b6cf4 100%);
    display: flex;
    position: relative;
    font-family: "Montserrat", sans-serif;
    color: white;
}

.Login:before {
    content: " ";
    position: absolute;
    height: 100%;
    width: 100%;
    background-image: url("./img/login-waves.svg");
    background-size: cover;
    z-index: 0;
}

.Login .column {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}

.Login h1 {
    font-size: 2.1rem;
    font-weight: bold;
    margin-top: 30px;
    text-align: center;
}

.Login h2 {
    font-size: 1.7rem;
    font-weight: bold;
    margin-bottom: 25px;
}

.Login .chooser {
    text-align: center;
    position: relative;
    z-index: 1;
}

.Login .accounts-container {
    border-radius: 6px;
    height: 335px;
    background-color: rgba(255, 255, 255, 0.5);
    overflow-y: auto;
    margin-bottom: 30px;
    padding: 40px;
    margin-left: 60px;
    margin-right: 60px;
}

.Login .node-info {
    text-align: center;
    margin-bottom: 60px;
}

.Login .node-info .name {
    font-size: 1.7rem;
    font-weight: bold;
}

.Login .node-info .peer-id {
    font-size: 0.9rem;
    font-weight: bold;
}

.Login .node-info .peer-id img {
    margin-right: 10px;
}

.Login .left-character {
    position: absolute;
    bottom: 0;
    left: 60px;
}

.Login .right-character {
    position: absolute;
    bottom: 0;
    right: 60px;
}

.Login .account {
    display: flex;
    color: #1b307a;
    padding-bottom: 55px;
    align-items: center;
}

.Login .account:last-child {
    padding-bottom: 0;
}

.Login .account .icon {
    height: 75px;
    width: 75px;
    border-radius: 36px;
    margin-left: 20px;
    margin-right: 20px;
    font-size: 45px;
    font-weight: bold;
    color: white;
    background-color: #3b6cf4;
}

.Login .account .name-address {
    text-align: left;
    font-weight: bold;
    height: 75px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}

.Login .account .name {
    font-size: 1.7rem;
}

.Login .account .address {
    font-size: 0.9rem;
}
