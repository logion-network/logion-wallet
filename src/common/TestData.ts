import { LegalOfficer, PostalAddress, UserIdentity } from '@logion/client';
import { Transaction } from '@logion/client/dist/TransactionClient.js';
import { Coin, CoinBalance, SYMBOL } from '@logion/node-api/dist/Balances.js';
import { PrefixedNumber, ATTO } from '@logion/node-api/dist/numbers.js';

import { ColorTheme, rgbaToHex } from './ColorTheme';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob
export const A_THIRD_LEGAL_OFFICER = "5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y"; // Charlie

export const DEFAULT_IDENTITY: UserIdentity = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@logion.network",
    phoneNumber: "+1234",
};

export const DEFAULT_ADDRESS: PostalAddress = {
    line1: "Place de le République Française, 10",
    line2: "boite 15",
    postalCode: "4000",
    city: "Liège",
    country: "Belgium",
};

export const COLOR_THEME: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: '#eff3fe',
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: '#d8e2fd',
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: 'eff3fe',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
        altBackground: '#eff3fe',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    recoveryItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondary: {
            background: '#eff3fe',
            foreground: '#000000',
        }
    },
    select: {
        background: '#ffffff',
        foreground: '#000000',
        menuBackgroundColor: '#ffffff',
        selectedOptionBackgroundColor: '#ffffff',
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: '#d8e2fd',
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    },
    dialog: {
        background: '#eff3fe',
        foreground: '#000000',
        borderColor: '#e6007a',
    },
    dialogTable: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
};

export const DEFAULT_COIN: Coin = {
    id: 'lgnt',
    iconId: 'lgnt',
    iconType: 'svg',
    name: "Logion",
    symbol: SYMBOL
};

export const DEFAULT_COIN_BALANCE: CoinBalance = {
    coin: DEFAULT_COIN,
    balance: new PrefixedNumber("42", ATTO),
    available: new PrefixedNumber("42", ATTO),
    level: 0.1,
};

export const DEFAULT_TRANSACTION: Transaction = {
    id: "id",
    from: "from",
    to: "to",
    pallet: "pallet",
    method: "method",
    transferValue: "420",
    tip: "0",
    fee: "125000",
    reserved: "0",
    total: "125420",
    createdOn: "2021-07-28T12:30:00.000",
    type: "Sent",
    transferDirection: "Sent",
    successful: true
}

export const DEFAULT_FAILED_TRANSACTION:Transaction = {
    ...DEFAULT_TRANSACTION,
    successful: false,
    error: {
        section: 'aSection',
        name: 'aName',
        details: 'someDetails'
    }
}

export const PATRICK: LegalOfficer = {
    name: "Patrick Gielen",
    address: DEFAULT_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "patrick@logion.network",
        phoneNumber: "+32 498 237 107",
    },
    postalAddress: {
        company: "MODERO",
        line1: "Huissier de Justice Etterbeek",
        line2: "Rue Beckers 17",
        postalCode: "1040",
        city: "Etterbeek",
        country: "Belgique"
    },
    node: "http://logion.patrick.com",
    logoUrl: "",
    nodeId: "",
};

export const GUILLAUME: LegalOfficer = {
    name: "Guillaume Grain",
    address: ANOTHER_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "g.grain@adrastee-lyon.fr",
        phoneNumber: "+33 4 78 52 87 56",
    },
    postalAddress: {
        company: "SELARL ADRASTEE",
        line1: "Gare des Brotteaux",
        line2: "14, place Jules Ferry",
        postalCode: "69006",
        city: "LYON",
        country: "France"
    },
    node: "http://logion.guillaume.com",
    logoUrl: "",
    nodeId: "",
};

export const ALAIN: LegalOfficer = {
    name: "Alain Barland",
    address: A_THIRD_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "alain.barland@auxilia-conseils.com",
        phoneNumber: "+33 2 48 67 50 50",
    },
    postalAddress: {
        company: "AUXILIA CONSEILS 18",
        line1: "Huissiers de Justice associés",
        line2: "7 rue Jean Francois Champollion Parc Comitec",
        postalCode: "18000",
        city: "Bourges",
        country: "France"
    },
    node: "http://logion.alain.com",
    logoUrl: "",
    nodeId: "",
};

export const legalOfficers: LegalOfficer[] = [
    PATRICK,
    GUILLAUME,
    ALAIN
];
