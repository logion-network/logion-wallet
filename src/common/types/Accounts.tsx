import { DateTime } from 'luxon';
import { InjectedAccount } from '@logion/extension';
import { LogionClient, Token } from '@logion/client';
import { LogionNodeApiClass, ValidAccountId } from '@logion/node-api';

export interface Account {
    readonly name: string,
    readonly accountId: ValidAccountId,
    readonly isLegalOfficer: boolean,
    readonly token?: Token,
}

export default interface Accounts {
    readonly all: Account[],
    readonly current?: Account,
}

export function buildAccounts(
    injectedAccounts: InjectedAccount[],
    userAddress: ValidAccountId | undefined,
    authenticatedClient: LogionClient,
    legalOfficers: Set<string>,
): Accounts {
    const all = injectedAccounts.map(injectedAccount => {
        const accountId = toValidAccountId(authenticatedClient.logionApi, injectedAccount);
        return {
            name: injectedAccount.meta.name!,
            accountId,
            isLegalOfficer: legalOfficers.has(injectedAccount.address),
            token: tokenOrUndefinedIfExpired(authenticatedClient.tokens.get(accountId)),
        };
    });

    const selectedAddress = currentOrDefaultAddress(injectedAccounts, userAddress, authenticatedClient);
    const current = all.find(account => account.accountId.toKey() === selectedAddress?.toKey());

    return {
        current,
        all,
    }
}

export function toValidAccountId(api: LogionNodeApiClass, injectedAccount: InjectedAccount): ValidAccountId {
    if(injectedAccount.type === "ethereum") {
        return api.queries.getValidAccountId(injectedAccount.address, "Ethereum");
    } else {
        return api.queries.getValidAccountId(injectedAccount.address, "Polkadot");
    }
}

function currentOrDefaultAddress(
    injectedAccounts: InjectedAccount[],
    currentAddress: ValidAccountId | undefined,
    authenticatedClient: LogionClient
): ValidAccountId | undefined {
    if(authenticatedClient === undefined) {
        return undefined;
    }
    const loggedAddresses = authenticatedClient.tokens.addresses;
    if(currentAddress !== undefined) {
        return currentAddress;
    } else if(loggedAddresses.length > 0) {
        const injectedAccountsKeys = injectedAccounts.map(injectedAccount => toValidAccountId(authenticatedClient.logionApi, injectedAccount).toKey());
        return loggedAddresses.find(account => injectedAccountsKeys.includes(account.toKey()));
    } else {
        return undefined;
    }
}

function tokenOrUndefinedIfExpired(token: Token | undefined): Token | undefined {
    if(token === undefined) {
        return undefined;
    }

    const expirationDateTime = token.expirationDateTime;
    if(expirationDateTime < DateTime.now()) {
        return undefined;
    } else {
        return token;
    }
}
