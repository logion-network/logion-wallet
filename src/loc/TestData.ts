import { LegalOfficerCase, UUID, Hash } from "@logion/node-api";
import { LocData, HashString } from "@logion/client";

export function buildLocRequest(locId: UUID, loc: LegalOfficerCase): LocData {
    return {
        ownerAddress: loc.owner,
        requesterAddress: loc.requesterAddress,
        requesterLocId: loc.requesterLocId,
        description: "Description",
        locType: loc.locType,
        createdOn: "",
        decisionOn: "",
        id: locId,
        status: loc.closed ? "CLOSED" : "OPEN",
        files: loc.files.map((locFile, index) => ({
            ...locFile,
            nature: HashString.fromValue(`File ${index}`),
            name: `File ${index}`,
            addedOn: "",
            published: false,
            restrictedDelivery: false,
            contentType: "text/plain",
            size: 42n,
            status: "DRAFT",
        })),
        metadata: loc.metadata.map((locFile, index) => ({
            ...locFile,
            name: HashString.fromValue(`Data ${index}`),
            value: HashString.fromValue(`Value ${index}`),
            addedOn: "",
            published: false,
            status: "DRAFT",
        })),
        links: loc.links.map((locFile, index) => ({
            ...locFile,
            name: `Link ${index}`,
            nature: HashString.fromValue(`Nature ${index}`),
            addedOn: "",
            target: locFile.id.toString(),
            published: false,
            status: "DRAFT",
        })),
        closed: false,
        verifiedIssuer: false,
        issuers: [],
    };
}
