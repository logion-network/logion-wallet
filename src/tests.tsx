import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
