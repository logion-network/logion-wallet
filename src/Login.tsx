import { ValidAccountId } from "@logion/node-api";
import { useState, useCallback } from 'react';
import { useNavigate, useLocation, Location } from 'react-router-dom';

import Container from 'react-bootstrap/Container';

import { useLogionChain } from './logion-chain';
import Button from './common/Button';
import Checkbox from './components/toggle/Checkbox';

import AbsoluteLogo from './AbsoluteLogo';

import './Login.css';
import Icon from "./common/Icon";

export const LOGIN_PATH = "/login";

export interface LocationState {
    referrer?: string;
}

function referrer(location: Location): string {
    const state = location.state as any;
    if(state && state.referrer) {
        return state.referrer;
    } else {
        return "/";
    }
}

export default function Login() {
    const location = useLocation();
    const navigate = useNavigate();
    const { connectedNodeMetadata, accounts, authenticate } = useLogionChain();
    const [ selectedAddresses, setSelectedAddresses ] = useState<ValidAccountId[]>([]);

    const startLogin = useCallback(async () => {
        await authenticate(selectedAddresses);
        navigate(referrer(location));
    }, [ selectedAddresses, navigate, location, authenticate ]);

    if(accounts === null || connectedNodeMetadata === null) {
        return null;
    }

    function selectAddress(address: ValidAccountId, selected: boolean) {
        let newSet;
        if(selected) {
            newSet = [ ...selectedAddresses ];
            newSet.push(address);
        } else {
            newSet = selectedAddresses.filter(item => item !== address);
        }
        setSelectedAddresses(newSet);
    }

    return (
        <div className="Login">
            <AbsoluteLogo />
            <Container className="column">
                <h1>
                    Please authenticate yourself by selecting<br />
                    one or several accounts below
                </h1>

                <div className="chooser">
                    <h2>The following accounts have been detected:</h2>

                    <div className="accounts-container">
                        {
                            accounts.all
                            .filter(address => address.token === undefined)
                            .map(address => (
                                <div className="account" key={ address.accountId.toKey() }>
                                    <Checkbox
                                        checked={ selectedAddresses.includes(address.accountId) }
                                        setChecked={ selected => selectAddress(address.accountId, selected) }
                                    />
                                    <div
                                        className="icon"
                                    >
                                        <Icon icon={{ id: address.accountId.type === "Ethereum" ? "metamask" : "polkadot-account" }} width="40px" />
                                    </div>
                                    <div
                                        className="name-address"
                                    >
                                        <div className="name">{ address.name }</div>
                                        <div className="address">{ address.accountId.address }</div>
                                    </div>
                                </div>
                            ))
                        }
                    </div>

                    <Button
                        disabled={ selectedAddresses.length === 0 }
                        onClick={ startLogin }
                    >
                        Log in
                    </Button>
                </div>

                <div className="node-info">
                    <p className="name">You are currently connected to the following logion node:</p>
                    <p className="peer-id">
                        <img src={process.env.PUBLIC_URL + "/assets/node.svg"} alt="node icon" />
                        Peer ID: {connectedNodeMetadata.peerId}
                    </p>
                </div>
            </Container>
            <div className="left-character">
                <img src={process.env.PUBLIC_URL + "/assets/login-left-character.svg"} alt="men pointing at box" />
            </div>
            <div className="right-character">
                <img src={process.env.PUBLIC_URL + "/assets/login-right-character.svg"} alt="women holding a loc" />
            </div>
        </div>
    );
}
