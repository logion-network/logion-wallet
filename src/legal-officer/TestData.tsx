import { DateTime } from 'luxon';
import { toIsoString } from "@logion/client";
import { ProtectionRequest } from '@logion/client/dist/RecoveryClient.js';

import { DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER } from "../common/TestData";

export const PENDING_PROTECTION_REQUESTS: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER.address,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER.address,
        decision: {
            rejectReason: null,
            decisionOn: null,
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(DateTime.fromISO('2021-06-10T11:40:00.000', {zone: "utc"})),
        isRecovery: false,
        addressToRecover: null,
        status: "PENDING"
    }
];

export const PROTECTION_REQUESTS_HISTORY: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER.address,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER.address,
        decision: {
                rejectReason: null,
                decisionOn: toIsoString(DateTime.fromISO('2021-06-10T11:40:00.000', {zone: "utc"})),
                locId: "locId"
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(DateTime.fromISO('2021-06-10T11:40:00.000', {zone: "utc"})),
        isRecovery: false,
        addressToRecover: null,
        status: "ACTIVATED"
    }
];

export const RECOVERY_REQUESTS_HISTORY: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER.address,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER.address,
        decision: {
            rejectReason: null,
            decisionOn: toIsoString(DateTime.fromISO('2021-06-10T11:40:00.000', {zone: "utc"})),
            locId: "locId"
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(DateTime.fromISO('2021-06-10T11:40:00.000', {zone: "utc"})),
        isRecovery: true,
        addressToRecover: "an-address",
        status: "PENDING"
    }
];
