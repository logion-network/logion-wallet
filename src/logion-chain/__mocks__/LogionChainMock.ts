import { DateTime } from "luxon";
import { LegalOfficer } from '@logion/client';
import { DEFAULT_LEGAL_OFFICER, legalOfficers } from 'src/common/TestData';
import { TEST_WALLET_USER } from 'src/wallet-user/TestData';
import { AxiosInstance } from 'axios';
import { Mock } from 'moq.ts';
import Accounts, { Account } from 'src/common/types/Accounts';
import { LogionClient } from '@logion/client/dist/LogionClient.js';
import { LogionClient as LogionClientMock } from '../../__mocks__/LogionClientMock';
import { api } from "src/__mocks__/LogionMock";
import { Call } from "../LogionChainContext";

export const LogionChainContextProvider = (props: any) => null;

export let clientMock: LogionClient | LogionClientMock | null = new LogionClientMock();

export let authenticate = jest.fn();

export let saveOfficer = jest.fn();

function getOfficer(address: string | undefined): LegalOfficer | undefined {
    return legalOfficers.find(legalOfficer => legalOfficer.address === address);
}

export let selectAddress = jest.fn();

export const DEFAULT_USER_ACCOUNT: Account = {
    name: "name",
    accountId: TEST_WALLET_USER,
    isLegalOfficer: false,
    token: {
        value: "token",
        expirationDateTime: DateTime.now(),
    },
}

export const DEFAULT_LEGAL_OFFICER_ACCOUNT: Account = {
    name: "name",
    accountId: DEFAULT_LEGAL_OFFICER,
    isLegalOfficer: true,
    token: {
        value: "token",
        expirationDateTime: DateTime.now(),
    },
}

export let accounts: Accounts | null = {
    current: DEFAULT_USER_ACCOUNT,
    all: [ DEFAULT_USER_ACCOUNT ]
}

export function setCurrentAddress(value: Account | undefined) {
    if(value === undefined) {
        accounts = {
            current: undefined,
            all: [ DEFAULT_USER_ACCOUNT ]
        };
    } else {
        accounts = {
            current: value,
            all: [ value ]
        };
    }
    if(context) {
        context.accounts = accounts;
    }
}

export function setAddresses(value: Accounts | null) {
    accounts = value;
}

export let axiosMock = new Mock<AxiosInstance>();

export function resetAxiosMock() {
    axiosMock = new Mock<AxiosInstance>();
}

let context: any = undefined;

export function setContextMock(value: any) {
    context = value;
}

let axiosFactory = () => axiosMock.object();

let signAndSend = jest.fn().mockResolvedValue(undefined);

let signer = {
    signAndSend,
};

export const NO_SUBMISSION: unknown = {
    canSubmit: () => true,
    isSuccessful: () => false,
    isError: () => false,
    error: null,
    result: null,
};

let extrinsicSubmissionState: unknown = NO_SUBMISSION;

export const SUCCESSFUL_SUBMISSION: unknown = {
    canSubmit: () => true,
    isSuccessful: () => true,
    isError: () => false,
    error: null,
    result: {
        status: {
            isFinalized: true,
        }
    },
};

export const FAILED_SUBMISSION: unknown = {
    canSubmit: () => false,
    isSuccessful: () => false,
    isError: () => true,
    error: "error",
    result: {
        status: {
            isFinalized: false,
        }
    },
};

export const PENDING_SUBMISSION: unknown = {
    canSubmit: () => false,
    isSuccessful: () => false,
    isError: () => false,
    error: null,
    result: {
        status: {
            isFinalized: false,
        }
    },
};

export function setExtrinsicSubmissionState(state: unknown) {
    extrinsicSubmissionState = state;
}

export function useLogionChain() {
    if(context) {
        return context;
    } else {
        return {
            injectedAccounts: null,
            api: api.object(),
            connectedNodeMetadata: {

            },
            selectAddress,
            accounts,
            axiosFactory,
            authenticate,
            getOfficer,
            saveOfficer,
            client: clientMock,
            isCurrentAuthenticated: () => true,
            authenticateAddress,
            signer,
            backendConfig: {
                [DEFAULT_LEGAL_OFFICER.address]: {
                    iDenfy: false,
                }
            },
            extrinsicSubmissionState,
            resetSubmissionState: () => {},
            submitSignAndSubmit: () => {},
            submitCall: (call: Call) => call(() => {}),
        };
    }
}

function authenticateAddress() {
    return clientMock;
}

export function setClientMock(mock: LogionClient | null) {
    clientMock = mock;
}
