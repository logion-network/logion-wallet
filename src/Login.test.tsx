import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DateTime } from 'luxon';

import Accounts, { Account } from './common/types/Accounts';
import { shallowRender } from './tests';
import { setAddresses, authenticate } from './logion-chain/__mocks__/LogionChainMock';
import { TEST_WALLET_USER } from './wallet-user/TestData';
import { DEFAULT_LEGAL_OFFICER } from "./common/TestData";
import Login from './Login';

jest.mock('./logion-chain');

const AUTHENTICATED_ADDRESS: Account = {
    name: "name authenticated",
    accountId: TEST_WALLET_USER,
    isLegalOfficer: false,
    token: {
        value: "token",
        expirationDateTime: DateTime.fromISO('2021-09-10T10:53:00.000'),
    },
};

const UNAUTHENTICATED_ADDRESS: Account = {
    name: "name unauthenticated",
    accountId: DEFAULT_LEGAL_OFFICER,
    isLegalOfficer: true,
};

const ADDRESSES: Accounts = {
    current: AUTHENTICATED_ADDRESS,
    all: [ UNAUTHENTICATED_ADDRESS ]
};

describe("Login", () => {

    it("renders", () => {
        setAddresses(ADDRESSES);
        const result = shallowRender(
            <Login />
        );
        expect(result).toMatchSnapshot();
    });

    it("enables log-in given unautenticated yet user", async () => {
        setAddresses(ADDRESSES);
        render(
            <Login />
        );

        await userEvent.click(screen.getByRole('checkbox'));
        await userEvent.click(screen.getByRole('button', {name: "Log in"}));

        await waitFor(() => expect(authenticate).toBeCalledWith([ DEFAULT_LEGAL_OFFICER ]));
    });
});
