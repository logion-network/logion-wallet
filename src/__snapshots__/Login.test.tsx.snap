// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders 1`] = `
<div
  className="Login"
>
  <AbsoluteLogo />
  <Container
    className="column"
    fluid={false}
  >
    <h1>
      Please authenticate yourself by selecting
      <br />
      one or several accounts below
    </h1>
    <div
      className="chooser"
    >
      <h2>
        The following accounts have been detected:
      </h2>
      <div
        className="accounts-container"
      >
        <div
          className="account"
        >
          <Checkbox
            checked={false}
            setChecked={[Function]}
          />
          <div
            className="icon"
          >
            <Icon
              icon={
                Object {
                  "id": "polkadot-account",
                }
              }
              width="40px"
            />
          </div>
          <div
            className="name-address"
          >
            <div
              className="name"
            >
              name unauthenticated
            </div>
            <div
              className="address"
            >
              vQx5kESPn8dWyX4KxMCKqUyCaWUwtui1isX6PVNcZh2Ghjitr
            </div>
          </div>
        </div>
      </div>
      <Button
        disabled={true}
        onClick={[Function]}
      >
        Log in
      </Button>
    </div>
    <div
      className="extensions-container"
    >
      <h2>
        You do not see your account?
      </h2>
      <Button
        className="metamask"
        disabled={true}
      >
        <Icon
          height="30px"
          icon={
            Object {
              "id": "metamask",
            }
          }
        />
         Try with MetaMask
      </Button>
    </div>
    <div
      className="node-info"
    >
      <p
        className="name"
      >
        You are currently connected to the following logion node:
      </p>
      <p
        className="peer-id"
      >
        <img
          alt="node icon"
          src="/assets/node.svg"
        />
        Peer ID: 
      </p>
    </div>
  </Container>
  <div
    className="left-character"
  >
    <img
      alt="men pointing at box"
      src="/assets/login-left-character.svg"
    />
  </div>
  <div
    className="right-character"
  >
    <img
      alt="women holding a loc"
      src="/assets/login-right-character.svg"
    />
  </div>
</div>
`;
